mbedtls/src/x509/mbedtls_wrappers.rs

```diff
@@ -0,0 +1,241 @@
+    }
+}
+
+impl Into<*mut x509_crt> for &MbedtlsBox<Certificate> {
+    fn into(self) -> *mut x509_crt {
+        self.inner.as_ptr() as *mut x509_crt
+    }
+}
+
+
+#[repr(transparent)]
+pub struct MbedtlsList<T> {
+    inner: *mut T,
+}
+
+#[cfg(feature = "threading")]
+unsafe impl<> Send for MbedtlsList<Certificate> {}
+
+#[cfg(feature = "threading")]
+unsafe impl<> Sync for MbedtlsList<Certificate> {}
+
+impl<> MbedtlsList<Certificate> {
+    pub fn new() -> Self {
+        Self { inner: core::ptr::null_mut() }
+    }
+
+    pub fn as_ref(&self) -> &Self {
+        self
+    }
+
+    pub fn empty(&self) -> bool {
+        self.inner == ::core::ptr::null_mut()
+    }
+    
+    pub fn push(&mut self, certificate: MbedtlsBox<Certificate>) -> () {
+        unsafe {
+            if self.inner == ::core::ptr::null_mut() {
+                self.inner = certificate.inner.as_ptr();
+            } else {
+                let mut cur : *mut x509_crt = self.inner as *mut x509_crt;
+                while (*cur).next != ::core::ptr::null_mut() {
+                    cur = (*cur).next;
+                }
+                (*cur).next = certificate.inner.as_ref().into();
+            }
+            core::mem::forget(certificate);
+        }
+    }
+
+    pub fn append(&mut self, list: MbedtlsList<Certificate>) -> () {
+        unsafe {
+            if self.inner == ::core::ptr::null_mut() {
+                self.inner = list.inner;
+            } else {
+                let mut cur : *mut x509_crt = self.inner as *mut x509_crt;
+                while (*cur).next != ::core::ptr::null_mut() {
+                    cur = (*cur).next;
+                }
+                (*cur).next = list.inner as *mut x509_crt;
+            }
+            core::mem::forget(list);
+        }
+    }
+
+    pub fn iter(&self) -> Iter<'_> {
+        Iter { next: self.inner as *const x509_crt, r: PhantomData }
+    }
+}
+
+impl<T> Drop for MbedtlsList<T> {
+    fn drop(&mut self) {
+        unsafe {
+            if self.inner != ::core::ptr::null_mut() {
+                drop_in_place(self.inner);
+                forward_mbedtls_free(self.inner as *mut c_void)
+            }
+        }
+    }
+}
+
+impl<> Clone for MbedtlsList<Certificate> {
+    fn clone(&self) -> MbedtlsList<Certificate> {
+        Certificate::from_cert(self.inner as *const x509_crt)
+    }
+}
+
+impl Into<*const x509_crt> for &MbedtlsList<Certificate> {
+    fn into(self) -> *const x509_crt {
+        self.inner as *const x509_crt
+    }
+}
+
+impl Into<*mut x509_crt> for &mut MbedtlsList<Certificate> {
+    fn into(self) -> *mut x509_crt {
+        self.inner as *mut x509_crt
+    }
+}
+
+impl Into<*mut x509_crt> for &MbedtlsList<Certificate> {
+    fn into(self) -> *mut x509_crt {
+        self.inner as *mut x509_crt
+    }
+}
+
+pub struct Iter<'a> {
+    next: *const x509_crt,
+    r: PhantomData<&'a x509_crt>,
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a Certificate;
```

Comment:
Make `Item = &'a MbedtlsBox<Certificate>` (also for IterMut). I think you can then define the struct as follows:
```rust
pub struct Iter<'a> {
    next: Option<&'a MbedtlsBox<Certificate>>,
}
```

Reply:
Since we need pointer to pointer in order to cast to the box it got a bit interesting, let me know if there are better options.